### PR TITLE
massive add of summary_texts

### PR DESCRIPTION
**Pull request by mmatera**
In this second stage, I add a "default" summary text to almost all the built-in classes. I did it "automatically" using a program that I list below. With this, I got a few cases that I had to correct by hand and ~140 cases that were not modified automatically.

```

import shutil
import mathics.builtin
import mpmath
import glob
import importlib
import pkgutil
import re
import os.path as osp
from mathics.settings import ENABLE_FILES_MODULE
from mathics.version import __version__  # noqa used in loading to check consistency.


from typing import List

from mathics.builtin.base import (
    Builtin,
    SympyObject,
    Operator,
    PatternObject,
)


builtins_directory = mathics.builtin.__file__[:-12]
__py_files__ = [
    osp.basename(f[0:-3])
    for f in glob.glob(osp.join(builtins_directory, "[a-z]*.py"))
]


def is_builtin(var):
    if var == Builtin:
        return True
    if hasattr(var, "__bases__"):
        return any(is_builtin(base) for base in var.__bases__)
    return False


# FIXME: redo using importlib since that is probably less fragile.
exclude_files = set(("codetables", "base"))
module_names = [
    f for f in __py_files__ #if re.match("^[a-z0-9]+$", f) if f not in exclude_files
]


disable_file_module_names = (
    [] if ENABLE_FILES_MODULE else ["files_io.files", "files_io.importexport"]
)


for subdir in (
    "arithfns",
    "atomic",
    "assignments",
    "box",
    "colors",
    "distance",
    "drawing",
    "files_io",
    "intfns",
    "list",
    "moments",
    "numbers",
    "specialfns",
    "string",
    "fileformats",
):
    import_name = f"mathics.builtin.{subdir}"

    if import_name in disable_file_module_names:
        continue

    builtin_module = importlib.import_module(import_name)
    for importer, modname, ispkg in pkgutil.iter_modules(builtin_module.__path__):
        module_names.append(f"{subdir}.{modname}")


def import_module(module_name: str):
    try:
        module = importlib.import_module(module_name)
    except Exception as e:
        print(e)
        print(f"    Not able to load {module_name}. Check your installation.")
        print(f"    mathics.builtin loads")
        return None

    if module:
        modules[module_name] = module

modules = dict()    
for module_name in module_names:
    print(module_name)
    import_module("mathics.builtin." + module_name)

print("\n\n", 80*"-", "\n\n")        

###################  Find the builtins for that summary_text is missing ###############
summary_text_by_module = dict()
count = 0

for module_name in modules:
    module_summary_texts = dict()
    module = modules[module_name]
    vars = dir(module)
    for name in vars:
        var = getattr(module, name)
        if (
            hasattr(var, "__module__")
            and var.__module__.startswith("mathics.builtin.")
            and var.__module__ != "mathics.builtin.base"
            and is_builtin(var)
            and not name.startswith("_")
            and var.__module__ == module.__name__
        ):  # nopep8

            instance = var(expression=False)

            if isinstance(instance, Builtin):
                # This set the default context for symbols in mathics.builtins
                if not hasattr(instance, "summary_text"):
                    module_summary_texts[var.__name__] = None
                    count = count + 1
    if module_summary_texts:
        summary_text_by_module[module_name] = module_summary_texts

#################  Read each module and modify the files to put a summary_text on each class ######


for name_of_module_to_fix in summary_text_by_module:
    module_to_fix = name_of_module_to_fix[16:]
    filename = builtins_directory +"/"+module_to_fix.replace(".", "/")+".py"
    builtins_to_fix = summary_text_by_module["mathics.builtin."+module_to_fix]
    cls_name = ""
    new_lines = []
    summary_ok = True
    with open(filename, "r") as f:
        lines = f.readlines()
        for line in lines:
            if line[:6]=="class ":
                cls_name = line[6:].split("(")[0]
                summary_ok =  (cls_name not in builtins_to_fix)
                new_lines.append(line)
                continue
            if line == line.strip():
                new_lines.append(line)
                continue
            strip_line = line[4:]
            if (not summary_ok) and  (strip_line[:4] == "def " or 
                                      strip_line.strip() == "pass" or
                                      strip_line.split("=")[0].strip() in 
                                      ("messages", "options", "rules")):
                new_lines.append(f'    summary_text = "{cls_name} summary still not available"\n')
                summary_ok = True
            new_lines.append(line)



    shutil.copyfile(filename, filename+"-old")
    with open(filename, "w") as f:
        for line in new_lines:
            f.write(line)
```

<a href="https://gitpod.io/#https://github.com/Mathics3/mathics-core/pull/223"></a>

### CODE REVIEW

**Review by rocky — CHANGES_REQUESTED** [review]
Let's hold off of this for another week. 

I was in the process of fully going over the last batch. If you look at the pattern that has gone on before your last change, summaries do not start with captial letters or have a final period. 

(There were other misspellings). 

I am in the process of trying to put out a docker release and having 67 changes a number of which say something like "not done yet" is a bit debatable.

**Comment by mmatera** [comment]
@rocky, the reason for doing this would be to add the sanity check in a way that new builtins need to have this field. The other way around would be to add the sanity check after completing all the ~2000 existing symbols and growing.

File: mathics/builtin/attributes.py [hunk 1]
```diff
@@ -204,7 +204,7 @@ class Protect(Builtin):
     """
 
     attributes = hold_all | protected
-    summary_text = "Protect summary still not available"
+    summary_text = "Set the attribute Protected."
```
Comment (TiagoCavalcante) [inline mathics/builtin/attributes.py:207]: Maybe add instead of set and removed instead of unset?

**Comment by rocky** [comment]
I have a couple of views on this. I am in awe of writing code or automating getting thing improved. Getting an idea of the size of the problem is useful as well. 

From a user side I am less thrilled. Being reminded hundreds of time that we haven't gotten our act together on for the docs isn't a great experience. In fact the more astute would have figured that out by virtue of the fact that there is no help text. 

And when I read lots text that looks like it was automatically generated, that is not a good sign either. In sum this can assist suggesting docstring but a human has to go over this and make sure it is right. 

If we want to keep note of places on the side for developers as an external file that is fine. Actually, I don't see why the test or extraction has to be in this code at all.  This could have been a totally separate project. So let's remove the code that was added inside core recently. In general, if things can be done outside of core, lets do them outside of core. 

Some other parts. When I look at the existing way summary text was added there are not capital letters or periods in the summary text in this codebase before.  I don't know whether that is a convention or `Information[]` "short" (`?`) option strings  versus `Information[]` "long" (`??`) option strings. But this is what came before.  Whatever we decide it should be consistent and this is not helping making things consistent. There are simple Python functions on strings for upper casing the first letter of word and removing periods at the end of a line.

**Comment by mmatera** [comment]
> I have a couple of views on this. I am in awe of writing code or automating getting things improved. Getting an idea of the size of the problem and having the code make suggestions is useful as well.
> 

Yep, the first idea here is to learn how to automatize certain massive transformations on the code. A more ambitious plan would be to globally change the format of the docstring documentation to something more standard, or to normalize the way in which properties are written in builtin classes. This is just a proof of concept.

> From a user side I am less thrilled. Being reminded hundreds of time that we haven't gotten our act together on for the docs isn't a great experience. In fact the more astute would have figured that out by virtue of the fact that there is no help text.
> 

This is a work in progress. The idea to put a standard text in all the ``summary_text`` was to ensure that all the classes have this field and that classes with a missing field be marked as something that requires intervention. My idea is that the majority of the summary texts could be automatically generated just by looking at certain properties of the class and the module (all the special functions that have a summary_text contains something like "the <classname> function", and for most of the constants it reads "the <classname> constant"). Also, it could be taken from the first lines of the docstring of the class.
 My plan is to have something like this before trying to merge this into master.



> And when I read lots text that looks like it was automatically generated, that is not a good sign either. In sum this can assist suggesting docstring but a human has to go over this and make sure it is right.
> 

This is relative: using an automatic text generator helps to get a uniform style. We can then pass the text through an automatic grammar checker to refine it, and finally, we can extract all the summary texts, make a table, and do a final human being check. 
This could save lots of hours of monotonous work.


> If we want to keep note of places on the side for developers as an external file that is fine. Actually, I don't see why the test or extraction has to be in this code at all. This could have been a totally separate project. So let's remove the code that was added inside core recently. In general, if things can be done outside of core, lets do them outside of core.
> 

I put the check code inside the "core" to force the future developers to fulfill certain standards in writing new builtins. However, probably the best would be to move this to a pytest check. 


> Some other parts. When I look at the existing way summary text was added there are not capital letters or periods in the summary text in this codebase before. I don't know whether that is a convention or `Information[]` "short" (`?`) option strings versus `Information[]` "long" (`??`) option strings. But this is what came before. Whatever we decide it should be consistent and this is not helping making things consistent. There are simple Python functions on strings for upper casing the first letter of word and removing periods at the end of a line.


To be honest, when I started with this I did it because you mention that this field had some relevance, but I didn't have clear where it was used. I guessed that it was a toolbar text to be used in the graphics interface. What I see now is that ``summary_text`` is the text part of the ``::usage`` message without the name of the symbol. This is why the string starts in lowercase. ``Information`` uses this string to show its "short" form, so we could have put this information in the "messages" dictionary. However, it is handy to have this as a separated property, so I would keep it, but rename it as "usage_text".

**Comment by rocky** [comment]
> 
> Yep, the first idea here is to learn how to automatize certain massive transformations on the code. A more ambitious plan would be to globally change the format of the docstring documentation to something more standard, or to normalize the way in which properties are written in builtin classes. This is just a proof of concept.

Ok - thanks for the explanation and it is a great proof of concept. But it reminds me of other aspect: the timing and the _way_ this was done which causes unnecessary confusion and stress.  

Yesterday, as I indicated the day before, my focus was on trying get out another docker image. It has been a month since the last one, a lot has happened, and there was reported breakage between the 4.0.0 tag and the last image. 

This is not the time to put in a PR that has many changes in it. In this situation, please leave the last minute kinds of changes before getting code out more publically to me (and I will try not to be abusive about it).  So #222 was really bad timing. Especially with something like "would like to merge this as soon as possible, to avoid creating merge conflic[t]s."  

To avoid merge conflicts don't contemplate massive changes to core right before something like a release. More generally, don't contemplate _any_ massive changes right before a release, or something that is going to take time to review whatever it is.

This PR was not marked as a draft until I marked it that way. And the code changes to core aside the class variable do not need to be in core. 

> 
> > From a user side I am less thrilled. Being reminded hundreds of time that we haven't gotten our act together on for the docs isn't a great experience. In fact the more astute would have figured that out by virtue of the fact that there is no help text.
> 
> This is a work in progress. The idea to put a standard text in all the `summary_text` was to ensure that all the classes have this field and that classes with a missing field be marked as something that requires intervention. My idea is that the majority of the summary texts could be automatically generated just by looking at certain properties of the class and the module (all the special functions that have a summary_text contains something like "the function", and for most of the constants it reads "the constant"). Also, it could be taken from the first lines of the docstring of the class. My plan is to have something like this before trying to merge this into master.


This is a good plan. But you should _communicate_ the plan _before_ starting on it. 

> 
> > And when I read lots text that looks like it was automatically generated, that is not a good sign either. In sum this can assist suggesting docstring but a human has to go over this and make sure it is right.

> 
> This is relative: using an automatic text generator helps to get a uniform style. We can then pass the text through an automatic grammar checker to refine it, and finally, we can extract all the summary texts, make a table, and do a final human being check. This could save lots of hours of monotonous work.

Having automated means of _checking_ is great. Automated means of suggesting is good too, but none of that code belongs in core as it currently is. We use black, and I used flake8 in my development environment. For the most part, you just don't see that this is going on in by adding code in the code base.

And then we have the problem that there was a PR that wasn't discussed and wasn't complete coupled with the fact that documentation hasn't been your strong point.  

So from my side after merging that in so I could get on with the other stuff I had _previosuly_ _planned_ now I am going over the grammar and spelling and style. The choice for me was do I let this slip, or do we let it go in and possibly correct it later. I am not happy about having to stop what I had planned and now consider this kind of choice. 

This was new stuff that I hadn't planned on doing and really didn't have to be done right then. And when I ask "please hold off for a week", I really mean "please hold off for a week". If totally privately you want to continue, that is okay. But keep it private. Let's not discuss this, this week. 


>
> I put the check code inside the "core" to force the future developers to fulfill certain standards in writing new builtins. However, probably the best would be to move this to a pytest check.

A great idea. Better is a pre-commit hook which is just checking new and changed code. 

While the _idea_ was great. It wasn't fully worked out.  It had bugs and any big idea is going to have bugs. 

>
> To be honest, when I started with this I did it because you mention that this field had some relevance, but I didn't have clear where it was used. I guessed that it was a toolbar text to be used in the graphics interface. What I see now is that `summary_text` is the text part of the `::usage` message without the name of the symbol. This is why the string starts in lowercase. `Information` uses this string to show its "short" form, so we could have put this information in the "messages" dictionary. However, it is handy to have this as a separated property, so I would keep it, but rename it as "usage_text".

Renaming to "usage_text" is a fine thing to do sometime after the weekend. The tutorial will need to be updated. We also need to add something in the developer guide about where adding new builtins functions get added to the code, which includes describing how things are organized (or will be fully organized when all of this is done).

**Comment by mmatera** [comment]
> 
> Ok - thanks for the explanation and it is a great proof of concept. But it reminds me of other aspect: the timing and the _way_ this was done which causes unnecessary confusion and stress.
> 

Sorry about that. I thought that this PR was about just a lot of innocuous changes that could help to improve the future additions to the base code. 


> Yesterday, as I indicated the day before, my focus was on trying get out another docker image. It has been a month since the last one, a lot has happened, and there was reported breakage between the 4.0.0 tag and the last image.
>

thank you for that. And sorry for distracting you from that task. In any case, I will not merge anything before consulting all of you.

 
> This is not the time to put in a PR that has many changes in it. In this situation, please leave the last minute kinds of changes before getting code out more publically to me (and I will try not to be abusive about it). So #222 was really bad timing. Especially with something like "would like to merge this as soon as possible, to avoid creating merge conflic[t]s."
> 

Yes, sorry again. I was not clear in what I meant by "as soon as possible". In any way, I meant "right now" or even "this month". The only rush on this is about how much we want to let grow the code without doing these "quality" checks. 

> To avoid merge conflicts don't contemplate massive changes to core right before something like a release. More generally, don't contemplate _any_ massive changes right before a release, or something that is going to take time to review whatever it is.
> 

OK, in that case, I am going to think about how to split this in smaller parts.


> This PR was not marked as a draft until I marked it that way. And the code changes to core aside the class variable do not need to be in core.
> 

Yes: sorry again for the rush. I just got excited about the aspects of how to implement the solution, and I forgot to mark this as a draft.

> > > From a user side I am less thrilled. Being reminded hundreds of time that we haven't gotten our act together on for the docs isn't a great experience. In fact the more astute would have figured that out by virtue of the fact that there is no help text.
> > 
> > 
> > This is a work in progress. The idea to put a standard text in all the `summary_text` was to ensure that all the classes have this field and that classes with a missing field be marked as something that requires intervention. My idea is that the majority of the summary texts could be automatically generated just by looking at certain properties of the class and the module (all the special functions that have a summary_text contains something like "the function", and for most of the constants it reads "the constant"). Also, it could be taken from the first lines of the docstring of the class. My plan is to have something like this before trying to merge this into master.
> 
> This is a good plan. But you should _communicate_ the plan _before_ starting on it.
> 

yep

> > > And when I read lots text that looks like it was automatically generated, that is not a good sign either. In sum this can assist suggesting docstring but a human has to go over this and make sure it is right.
>

I am improving that now. For the next week, I hope to have a more acceptable version of almost all the summary_texts
 
> > This is relative: using an automatic text generator helps to get a uniform style. We can then pass the text through an automatic grammar checker to refine it, and finally, we can extract all the summary texts, make a table, and do a final human being check. This could save lots of hours of monotonous work.
> 
> Having automated means of _checking_ is great. Automated means of suggesting is good too, but none of that code belongs in core as it currently is. We use black, and I used flake8 in my development environment. For the most part, you just don't see that this is going on in by adding code in the code base.
> 
> And then we have the problem that there was a PR that wasn't discussed and wasn't complete coupled with the fact that documentation hasn't been your strong point.
> 
> So from my side after merging that in so I could get on with the other stuff I had _previously_ _planned_ now I am going over the grammar and spelling and style. The choice for me was do I let this slip, or do we let it go in and possibly correct it later. I am not happy about having to stop what I had planned and now consider this kind of choice.
> 

Again, I didn't expect that you review all the changes in the documentation, this weekend or this year. What I am going to do is to 
pass the new lines through a spelling and grammar checker, read them, and try to leave something that slowly can be improved. 

> This was new stuff that I hadn't planned on doing and really didn't have to be done right then. And when I ask "please hold off for a week", I really mean "please hold off for a week". If totally privately you want to continue, that is okay. But keep it private. Let's not discuss this, this week.
> 


> > I put the check code inside the "core" to force the future developers to fulfill certain standards in writing new builtins. However, probably the best would be to move this to a pytest check.
> 
> A great idea. Better is a pre-commit hook which is just checking new and changed code.
> 
> While the _idea_ was great. It wasn't fully worked out. It had bugs and any big idea is going to have bugs.
> 
> > To be honest, when I started with this I did it because you mention that this field had some relevance, but I didn't have clear where it was used. I guessed that it was a toolbar text to be used in the graphics interface. What I see now is that `summary_text` is the text part of the `::usage` message without the name of the symbol. This is why the string starts in lowercase. `Information` uses this string to show its "short" form, so we could have put this information in the "messages" dictionary. However, it is handy to have this as a separated property, so I would keep it, but rename it as "usage_text".
> 
> Renaming to "usage_text" is a fine thing to do sometime after the weekend. The tutorial will need to be updated. We also need to add something in the developer guide about where adding new builtins functions get added to the code, which includes describing how things are organized (or will be fully organized when all of this is done).

OK, so I am going to stop pushing things here, and complete the next steps locally.

**Comment by mmatera** [comment]
Well, there is my of describing all this mess:
https://github.com/Mathics3/mathics-core/discussions/236#discussion-3954793

**Pull request closed without merging**